src/target/riscv/riscv-011.c

```diff
@@ -455,16 +455,24 @@ static uint64_t dbus_read(struct target *target, uint16_t address)
 	dbus_status_t status;
 	uint16_t address_in;
 
-	unsigned i = 0;
-	do {
-		status = dbus_scan(target, &address_in, &value, DBUS_OP_READ, address, 0);
+	/* Assume dbus is idle and there aren't any busy/error states pending.
+	 * If there happens to be a busy state pending, we will clear it after we
+	 * discover it as part of the NOP scan. */
+	for (unsigned i = 0; i < 256; i++) {
+		dbus_scan(target, NULL, NULL, DBUS_OP_READ, address, 0);
```

Comment:
I'm sorry, this first scan can absolutely encounter busy. The next scan can *create* the busy condition (v11 doesn't specify that DBUS_OP_NOP isn't allowed to generate a BUSY result. In fact I think DBUS_OP_NOP just does a read as well). If the first scan encounters busy, the DTM doesn't send anything over DMI. Then the DBUS_OP_NOP needlessly is sent to see if you got the result. Then misleadingly you could interpret the result of a DBUS_OP_NOP as the result of the DBUS_OP_READ. Probably this happens to work, but it's weird.

Here is what I mean. 
Let's say you first read from FOO, and then read from BAR.

Here's the FOO read:
```
dbus_scan(DBUS_OP_READ, FOO)
   -> busy=0, as recorded in CAPTURE_DR.
   -> during UPDATE_DR, the READ is sent across DMI. This corresponds to the fact that busy=0.
...
   -> read completes on DMI.
dbus_scan(DBUS_OP_NOP, FOO)
   -> since the read completed, CAPTURE_DR gets busy = 0 and the read data, and address = FOO.
   -> Update DR *still sends the NOP operation into DMI* This can result in busy for the next operation!
```
Since busy was 0, address = FOO, this read was correctly interpreted as having finished. So you break out of the loop.

Now you want to read from BAR
*Let's say you do this really fast and the second FOO DMI operation has not yet completed.*

```
dbus_scan(DBUS_OP_READ, BAR)
   -> result *is* busy=1, as recorded in CAPTURE_DR, because the FOO NOP has not completed.
   -> during UPDATE_DR, the READ is NOT sent across DMI. This corresponds to the reported busy=1.
...
   -> FOO NOP finally completes on DMI.
dbus_scan(DBUS_OP_NOP, BAR)
   -> since the *NOP* completed, CAPTURE_DR gets busy = 0. But, busy is sticky so the reported busy=1 since you haven't cleared it yet. (It took me a while to realize that the stickyness of busy was the only reason this made any real sense).
   -> Update DR does nothing because you haven't cleared the busy yet.
```
You observe that busy = 1 so you finally clear it and go back through the loop again. This seems... convoluted. It costs nothing to check the first busy result and clear it when it first affects you. I don't understand, why not do it?